### ISSUE TITLE
Watch not working on windows

### ISSUE THREAD

**Issue by spectras** [jlmakes/karma-rollup-preprocessor#65]
Using version 7.0.3.
Past the initial preprocessing, no file change ever trigger test again.

I traced the issue to the matching in the watcher. It passes down changed file path to Karma's fileList. In turn, the fileList does this:
```js
//karma\lib\file-list.js:48
_findFile (path, pattern) {
  if (!path || !pattern) return
  return this._getFilesByPattern(pattern.pattern).find((file) => file.originalPath === path)
}
```
The issue is:
* `path` is a windows path, in my case, `D:\home\projects\skeleton\skeleton-lib\tests\index.js`.
* `file.originalPath` is not: `'D:/home/projects/skeleton/skeleton-lib/tests/index.js'`.

It looks like Karma's `fileList` expects the given path to always be posix format, and `karma-rollup-preprocessor` sends windows format instead.

**Comment by spectras** [comment]
**Note:** a quick-and-dirty change in `karma-rollup-preprocessor/lib/Watcher.js` like this:
```diff
-		this.emitter._fileList.changeFile(path, true)
+		this.emitter._fileList.changeFile(path.replace(/\\/g, '/'), true)
```
…makes it work on my system.

**Comment by jlmakes** [comment]
Something like this?

```js
const sep = require('path').sep

//...

this.emitter._fileList.changeFile(sep !== '/' ? path.replace(/\\/g, '/') : path, true)
```

**Comment by Lalem001** [comment]
@jlmakes [path.normalize](https://nodejs.org/api/path.html#path_path_normalize_path) might work here. Not at my computer, so cannot double check.

**Comment by Lalem001** [comment]
Took a look, and my above comment with `path.normalize` will not work.

Take a look instead at [slash](https://github.com/sindresorhus/slash).

**Comment by jlmakes** [comment]
It's basically a wrapper around the same string replace:

```js
// slash.js

'use strict';
module.exports = path => {
	const isExtendedLengthPath = /^\\\\\?\\/.test(path);
	const hasNonAscii = /[^\u0000-\u0080]+/.test(path);

	if (isExtendedLengthPath || hasNonAscii) {
		return path;
	}

	return path.replace(/\\/g, '/');
};
```

I understand testing for ASCII, but I had to search for "extended length paths":
https://stackoverflow.com/questions/21194530/what-does-mean-when-prepended-to-a-file-path

**Comment by Lalem001** [comment]
I know, was just thinking that it could potentially handle the edge cases for you if they ever came up.

**Comment by jlmakes** [comment]
I'm generally cautious of adding dependencies, but I'm not opposed to it. I just don't fully understand if those cases are overly defensive for Karma and Rollup. As in, will non-ASCII or extended length paths occur in a JavaScript code base, or the build process therein?

> That would however break projects that actually use backslashes in file names on posix systems. — @spectras 

Similarly, this would still need to be addressed in the implementation. So, something like this?

```js
const slash = require('slash')
const sep = require('path').sep

//...

this.emitter._fileList.changeFile(sep !== '/' ? slash(path) : path, true)
```

**Comment by jlmakes** [comment]
I'm generally cautious of adding dependencies, but I'm not opposed to it. I just don't fully understand if those cases are overly defensive for Karma and Rollup. As in, will non-ASCII or extended length paths occur in a JavaScript code base, or the build process therein?

> That would however break projects that actually use backslashes in file names on posix systems. — @spectras 

This would still need to be addressed in the implementation. So, something like this?

```js
const slash = require('slash')
const sep = require('path').sep

//...

this.emitter._fileList.changeFile(sep !== '/' ? slash(path) : path, true)
```

**Comment by Lalem001** [comment]
Just want to point out that `\u0000` does show up when certain rollup plugins are involved.

**Comment by jlmakes** [comment]
I pushed an optimistic "fix" to `development`. Hoping @spectras can help test.

**Comment by spectras** [comment]
Good news and bad news!

**Bad news:** 5668d68 does not work here.

**Good news:** the fix in it works.
- After seeing 5668d68 did not work, I tried 7.0.4, and I have the same problem in it.
- So I downgraded to 7.0.3 and backported the commit. This worked.

So, the fix works, but 7.0.4 introduced another issue. I'll open a separate topic.
